### ISSUE TITLE
您好，当我下载代码并经过catkin_make以后，报了Could not find a package configuration file provided by   "sensor_fusion_comm" with any of the following names:      sensor_fusion_commConfig.cmake     sensor_fusion_comm-config.cmake的错误，请问怎么解决

### ISSUE THREAD

**Issue by bryantaoli** [milkytipo/MSF_developed#2]
您好，当我下载代码并经过catkin_make以后，报了Could not find a package configuration file provided by   "sensor_fusion_comm" with any of the following names:      sensor_fusion_commConfig.cmake     sensor_fusion_comm-config.cmake的错误，请问怎么解决？
![image](https://user-images.githubusercontent.com/30526360/78007502-f6100d80-7370-11ea-8f93-fb227937caea.png)

**Comment by milkytipo** [comment]
这个MSF_developed是基于原版msf的，其实原理就是把原来的ethzasl_msf当做一个库来使用，框架是用的原来的。所以你需要把原来的msf也装上，我做了一些bug的修复，你可以直接用我改好的：https://github.com/milkytipo/ethzasl_msf

**Comment by Ynjxsjmh** [comment]
@milkytipo Where to place original MSF?

**Comment by milkytipo** [comment]
> [@milkytipo](https://github.com/milkytipo) Where to place original MSF?

place at the same dir, or other dir and add the path in the configuration

**Comment by Ynjxsjmh** [comment]
@milkytipo Thank you, I follow [this link](https://www.liuxiao.org/2016/07/ros-%E5%A4%9A%E4%BC%A0%E6%84%9F%E5%99%A8%E5%8D%A1%E5%B0%94%E6%9B%BC%E8%9E%8D%E5%90%88%E6%A1%86%E6%9E%B6-ethzasl-msf-framework-%E7%BC%96%E8%AF%91%E4%B8%8E%E4%BD%BF%E7%94%A8/), and the final dir structure looks like

```
$ tree
.
└── src
    ├── glog_catkin
    ├── catkin_simple
    ├── asctec_mav_framework
    ├── ethzasl_msf
    └── MSF_developed
```

Besides this, I need to manually cmake, make and make install GeographicLib which is required by read_dataset

https://github.com/milkytipo/MSF_developed/blob/da6ec28ade3736e8a807c90cc36b9b199fa4b082/src/read_dataset/main.cpp#L19-L20

**Comment by Ynjxsjmh** [comment]
@milkytipo Hi Zida,


What should I modify in ORB SLAM `ros_stereo.cc` to advertise `/slam/tf` topic? I tried following code, but the result is ‌weird‌.

```
void ImageGrabber::GrabStereo(const sensor_msgs::ImageConstPtr& msgLeft,const sensor_msgs::ImageConstPtr& msgRight)
{
    ......

    cv::Mat Tcw;
    if(do_rectify)
    {
        cv::Mat imLeft, imRight;
        cv::remap(cv_ptrLeft->image,imLeft,M1l,M2l,cv::INTER_LINEAR);
        cv::remap(cv_ptrRight->image,imRight,M1r,M2r,cv::INTER_LINEAR);
        Tcw = mpSLAM->TrackStereo(imLeft,imRight,cv_ptrLeft->header.stamp.toSec());
    }
    else
    {
        Tcw = mpSLAM->TrackStereo(cv_ptrLeft->image,cv_ptrRight->image,cv_ptrLeft->header.stamp.toSec());
    }

    geometry_msgs::TransformStamped transformStamped;
    transformStamped.header.stamp = msgLeft->header.stamp;
    transformStamped.header.frame_id = "map";
    transformStamped.child_frame_id = "camera";

    transformStamped.transform.translation.x = Tcw.at<float>(0, 3);
    transformStamped.transform.translation.y = Tcw.at<float>(1, 3);
    transformStamped.transform.translation.z = Tcw.at<float>(2, 3);

    Eigen::Matrix3d R;
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            R(i, j) = Tcw.at<float>(i, j);
        }
    }

    Eigen::Quaterniond q(R);
    transformStamped.transform.rotation.x = q.x();
    transformStamped.transform.rotation.y = q.y();
    transformStamped.transform.rotation.z = q.z();
    transformStamped.transform.rotation.w = q.w();

    tf_pub.publish(transformStamped);
}
```

Why is the SLAM result vertical to the GPS and MSF result?

When I run your [plotTrajectory](https://github.com/milkytipo/plotTrajectory) example, and uncomment 

https://github.com/milkytipo/plotTrajectory/blob/45055cfa64ba647bbe6f515239bbf708840fe8cc/plotTrajectory.py#L80

The SLAM result is vertical to the other result, is this common?

![Image](https://github.com/user-attachments/assets/46cd75dd-baba-443f-85d3-316328113b19)

**Comment by milkytipo** [comment]
> [@milkytipo](https://github.com/milkytipo) Hi Zida,
> 
> What should I modify in ORB SLAM `ros_stereo.cc` to advertise `/slam/tf` topic? I tried following code, but the result is ‌weird‌.
> 
> ```
> void ImageGrabber::GrabStereo(const sensor_msgs::ImageConstPtr& msgLeft,const sensor_msgs::ImageConstPtr& msgRight)
> {
>     ......
> 
>     cv::Mat Tcw;
>     if(do_rectify)
>     {
>         cv::Mat imLeft, imRight;
>         cv::remap(cv_ptrLeft->image,imLeft,M1l,M2l,cv::INTER_LINEAR);
>         cv::remap(cv_ptrRight->image,imRight,M1r,M2r,cv::INTER_LINEAR);
>         Tcw = mpSLAM->TrackStereo(imLeft,imRight,cv_ptrLeft->header.stamp.toSec());
>     }
>     else
>     {
>         Tcw = mpSLAM->TrackStereo(cv_ptrLeft->image,cv_ptrRight->image,cv_ptrLeft->header.stamp.toSec());
>     }
> 
>     geometry_msgs::TransformStamped transformStamped;
>     transformStamped.header.stamp = msgLeft->header.stamp;
>     transformStamped.header.frame_id = "map";
>     transformStamped.child_frame_id = "camera";
> 
>     transformStamped.transform.translation.x = Tcw.at<float>(0, 3);
>     transformStamped.transform.translation.y = Tcw.at<float>(1, 3);
>     transformStamped.transform.translation.z = Tcw.at<float>(2, 3);
> 
>     Eigen::Matrix3d R;
>     for (int i = 0; i < 3; ++i) {
>         for (int j = 0; j < 3; ++j) {
>             R(i, j) = Tcw.at<float>(i, j);
>         }
>     }
> 
>     Eigen::Quaterniond q(R);
>     transformStamped.transform.rotation.x = q.x();
>     transformStamped.transform.rotation.y = q.y();
>     transformStamped.transform.rotation.z = q.z();
>     transformStamped.transform.rotation.w = q.w();
> 
>     tf_pub.publish(transformStamped);
> }
> ```
> 
> Why is the SLAM result vertical to the GPS and MSF result?
> 
> When I run your [plotTrajectory](https://github.com/milkytipo/plotTrajectory) example, and uncomment
> 
> https://github.com/milkytipo/plotTrajectory/blob/45055cfa64ba647bbe6f515239bbf708840fe8cc/plotTrajectory.py#L80
> 
> The SLAM result is vertical to the other result, is this common?
> 
> ![Image](https://github.com/user-attachments/assets/46cd75dd-baba-443f-85d3-316328113b19)

yes, it is vertical. I remember this one. This is common as your understanding. You already get what it is!

I am sure you can find a way to change the direction, sorry i cannot remember how to change it because it was many years ago  :(

**Comment by Ynjxsjmh** [comment]
To summarize what I have done,

1. Modify `ros_stereo.cc` to publish tf, I use tf2 with following CUDA version which also requires modification in `pose_sensorhandler.hpp::MeasurementCallback()`

https://github.com/thien94/ORB_SLAM2_CUDA/blob/4a6156cf75b22630a6d5eceee590a2b29ab48452/Examples/ROS/ORB_SLAM2_CUDA/lib/SlamData.cc#L74-L90 

2. Compile ORB_SLAM2 with `build_ros.sh` to get `Stereo` executable file.

The problems I need to solve:

1. (SLAM, GT) and (MSF, GPS) results are vertical. 
   > SLAM的轨迹图是通过旋转矩阵将SLAM本体坐标系下的轨迹映射到ENU系下的结果。

   From the quote from your thesis paper, I guess the problem might be that I don't convert the coordinate.

2. GT file timestamp couldn't match the saved bag file, no matter the bag file in [MSF_developed](https://github.com/milkytipo/MSF_developed/tree/master/MSF_noised_results) or in [plotTrajectory](https://github.com/milkytipo/plotTrajectory).

```
$ evo_ape tum kitti_00_groudtruth.tum pose.tum -v -a -p
--------------------------------------------------------------------------------
Loaded 4540 stamps and poses from: kitti_00_groudtruth.tum
Loaded 46638 stamps and poses from: pose.tum
Synchronizing trajectories...
[ERROR] found no matching timestamps between kitti_00_groudtruth.tum and pose.tum with max. time diff 0.01 (s) and time offset 0.0 (s)
```
